### ISSUE TITLE
Replace submodules with subtrees

### ISSUE THREAD

**Issue by oli-obk** [rust-lang/rust#70651]
cc @rust-lang/clippy @RalfJung 

cc @rust-lang/compiler @rust-lang/infra  we gate on clippy soon, so if you break clippy you need to fix it immediately in the same PR. This is trivial now though, since we are basically in a monorepo state, so you just edit stuff in `src/tools/clippy/` until everything works again.

Synchronizations of the remote repo and the local subtree is explained in CONTRIBUTING.md

The first port (clippy) is being done in https://github.com/rust-lang/rust/pull/70655

**Comment by RalfJung** [comment]
Cc @rust-lang/miri because we have that now :D

**Comment by eddyb** [comment]
Hmm, I was led to believe that `git subrepo` was nesting `git` repositories, in a way which `git` natively handles. But it turns out it's just a monorepo sync tool, same as manually making large commits. I don't think losing commit history is acceptable.

At a glance, `git subtree` looks like it might do the right thing, i.e. keep the repository nesting of a submodule, but I haven't tried it yet to see how it actually behaves in practice.

**Comment by oli-obk** [comment]
I moved to git subtree.

**Comment by eddyb** [comment]
In case it gets lost on the PR, I said: https://github.com/rust-lang/rust/pull/70655#issuecomment-607240782
> Whoa, clippy commit hashes were kept intact, this makes me really happy!11

**Comment by RalfJung** [comment]
> I moved to git subtree.

Would be good to get the pro's and con's of that decision documented somewhere. `git submodule` advertises itself as a better alternative to `git subtree`; why is `subtree` still the better choice for us?

**Comment by oli-obk** [comment]
I don't know why `git subrepo` is better than `git subtree` other than what the documentation of `git subrepo` states at https://github.com/ingydotnet/git-subrepo#benefits

> Fixes known rebase failures with git-subtree.

Git subtree is better than `git subrepo because it preserves all commits in history instead of squishing them in the name of the person who synchronizes it.

**Comment by eddyb** [comment]
`git subtree` is upstream, which is nice, and it also supports preserving the commits better.
I believe the "rebase failures" that `git subrepo` talks about are actually an anti-pattern: rebasing a merge commit, it's not meant to work *at all*.

`git subrepo` seems to want to just throw all the changes into single commits so you "don't have to deal with it", whereas `git subtree` is more history-preserving-oriented, which I prefer.

**Comment by RalfJung** [comment]
> git subtree is better than git subrepo because it preserves all commits in history instead of squishing them in the name of the person who synchronizes it.

Is this true in both directions (subtree pull/push)?

Does `git subtree` also have the problem that a "subtree push" needs a commit on the rustc side?

**Comment by oli-obk** [comment]
no, `git subtree push` does not need a commit on the rustc side.

The only downside of `git subtree` is that you need to specify the folder name and target repo on every synchronization (because there's no synchronization metadata file)

**Comment by Manishearth** [comment]
I would prefer the history preserving option. I was under the impression that subrepo had a no squash mode but that could be subtree.

**Comment by RalfJung** [comment]
> The only downside of git subtree is that you need to specify the folder name and target repo on every synchronization (because there's no synchronization metadata file)

This still needs to be reflected in https://github.com/rust-lang/rust/pull/70654 then.
I presume we could have script for that in the rustc repo?

**Comment by oli-obk** [comment]
> This still needs to be reflected in #70654 then.

it already is:

> You always need to specifiy the `-P` prefix to the subtree directory. If you specify the wrong directory
you'll get very fun merges that try to push the wrong directory to the remote repository. Luckily you
can just abort this without any consequences.

**Comment by oli-obk** [comment]
>  I was under the impression that subrepo had a no squash mode but that could be subtree.

subtree doesn't squash by default and has a sqash option.

**Comment by flip1995** [comment]
@oli-obk IIUC, Clippy can then fix a nightly version and doesn't have to use the rustc master anymore (correct?). If that's the case, I'd go ahead and make these changes to the Clippy CI/infra, so it will be ready once the `subtree` PRs are ready to get merged.

**Comment by RalfJung** [comment]
> it already is:

It doesn't say that you need to give the target repo (and the example commands do not mention a target repo, just a folder name).

**Comment by oli-obk** [comment]
> Clippy can then fix a nightly version and doesn't have to use the rustc master anymore (correct?). If that's the case, I'd go ahead and make these changes to the Clippy CI/infra, so it will be ready once the subtree PRs are ready to get merged.

well, only if we make the synchronization at specific commits that work on the current nightly. Between two nightlies there can be many states where the clippy subtree will only compile with a master-rustc.

**Comment by flip1995** [comment]
IMO we want to do this, since otherwise, we would have to do rustup commits in Clippy just as often as before. It will just be easier, since it will only be a `git subtree push`.

**Comment by oli-obk** [comment]
Well, we can do it at random points in time, but yes, we can target those points in time to the commit that caused the most recent nightly to be created.

**Comment by Manishearth** [comment]
Right, we can have a nightly clippy CI job that notices when clippy stops building with master, and sync then.

**Comment by oli-obk** [comment]
at that point we can just have a nightly job that synchronizes once a night automatically to ensure that even *nonbreaking* changes get synchronized as often as possible.

**Comment by eddyb** [comment]
@oli-obk @RalfJung What do you think of this logic? Just thought of it, not sure where to leave it.

`git subtree pull` is an *injection*, `git subtree push` is a *projection*, hence why pulling preserves commits verbatim (hash included) while pushing rewrites them to keep only the relevant parts. They're both fully information preserving for the subtree directory, and *only* for that directory.

**Comment by RalfJung** [comment]
> git subtree pull is an injection, git subtree push is a projection

I have no idea what this means in this context.^^
I mean I know waht both do as you explained it elsewhere, but this terminology here is not useful at all I feel.

**Comment by tmandry** [comment]
I'm really glad this effort is underway. On Fuchsia we try to roll to a new tip-of-tree rust compiler every few weeks, which helps with catching bugs. If we could we'd roll more often. But the main thing slowing us down is the constant state of toolstate breakage.

Some of this (like rustfmt) we can work around in practice by distributing separate versions, but it's a pain to set up and maintain. Some tools (like clippy and rls) really have to match the version of the compiler.

This seems like it would make toolstate breakage a thing of the past, which is why I'm very supportive of this effort and would be glad to help however I can.

**Comment by oli-obk** [comment]
We've [protyped this with clippy](https://github.com/rust-lang/rust-clippy/issues/5565) but are waiting for an [upstream fix to `git subtree`](https://github.com/gitgitgadget/git/pull/493) before it becomes feasible for the rustc repo without a patched `git`

**Comment by Manishearth** [comment]
Yeah, I'm hoping for that fix to land before I start seriously pushing it for others. I wanna start having the conversations soon though

**Comment by phil-blain** [comment]
Hi! I contribute to Git in my spare-time and I'm really interested in improving the user experience around submodules. I would really be interested to know the reasons for why this change of approach is being made for your project :) After all, `git subtree` is in `contrib/`, so it's not an "official" Git command and it's not being actively developed...

**Comment by oli-obk** [comment]
I summarized the reasons in https://github.com/rust-lang/compiler-team/issues/266

TLDR: keeping submodules in sync is a lot of effort, we need to be able to easily make changes to the "sub-whatever" to keep it from breaking due to other changes in the main repo, but at the same time not have a pure monorepo, because the "sub-whatever" is also being developed out of tree.

**Comment by Manishearth** [comment]
@phil-blain The links Oli posted should give some context, but in short:

  - We have multiple subprojects that are logically separate projects, but have strong code dependencies with this repository. Mainly, they make use of internal, unstable APIs that are prone to a lot of change.
 - If these projects are vendored in tree, this largely just works out: When an internal API is changed, the person changing it will update all internal consumers, as usual. However, the subproject needs to do all development on this main repo, which can slow down stuff.
 - With a submodule, this is far more annoyign to do and would make it harder to iterate on internal APIs. Instead of making compiler developers need to make frequent coordinated submodule updates (which would slow down compiler development and push compiler developers into situations where they may prefer to not update internal APIs even though that's the technically best option), we would have CI catch submodule breakages and submodule maintainers would update things. This made life easier for compiler devs but submodule devs, who may not be 100% familiar with the changes to the internal API, would have to figure this out, and it would often be a lot of work.
 - With subtrees, we have the best of both worlds: Subproject development is done in a separate repo, but whenever someone breaks an internal API it is very easy to fix up the subprojects. We do biweekly syncs and it all works pretty great.

While I do think the UX around submodules could use a lot of love, in this case I think it's more of them not being the right tool for the job.

**Comment by oli-obk** [comment]
TLDR: we could create a `git-subtree` command installable via `cargo install rustc-git-subtree` to make the use of `git subtree` on the rustc repo easy.

Some links to start out with:

* [clippy's tracking issue for the subtree sync problems](https://github.com/rust-lang/rust-clippy/issues/5565)
* [there's desire to move the `doc` directory to a subtree](https://github.com/rust-lang/rust/issues/76953)
* [recent discussion on zulip](https://rust-lang.zulipchat.com/#narrow/stream/131828-t-compiler/topic/subtree/near/214246408)

Even if the git subtree patch is merged and distributed, it may take years to appear in everyone's distro.
@Mark-Simulacrum suggested that we could remove the blocker on the git subtree patch by distributing it ourselves. All we need is a version that works for us after all. So basically we create a new rust project in the rust-lang org and all it does when installed is invoke the scripts (that we'll shamelessly-but-with-the-right-license copy from the git-PR). This way anyone who wants do subtree synchronizations can "overwrite" their `git subtree` command, by installing ours via `cargo install rustc-git-subtree`. This will put the `git-subtree` binary into the `PATH` like it does with `cargo-*` subcommands. This will take precedence over the `git-subtree` of your OS, and thus just work out of the box.

**Comment by phil-blain** [comment]
> This will put the git-subtree binary into the PATH like it does with cargo-* subcommands. This will take precedence over the git-subtree of your OS, and thus just work out of the box.

Careful though, people running `git-subtree` will get your version, but people using `git subtree` will get the original script from `contrib/` in git.git, which gets installed to `git --exec-path` by most distros' git package, since they install subtree by default.

**Comment by Mark-Simulacrum** [comment]
Probably we want to call it rustc-subtree then, or something like that. Doesn't seem like a huge concern -- people interacting with it are a relatively small set, hopefully.

**Comment by flip1995** [comment]
This is a good idea, IMO. Some things the `rustc-subtree` command should take care of:

- download and use the new fixed script
- set the stack limit higher (on command line this is done with `ulimit -s 60000`) for large subrepos on first use.
- nicer command line features, either configurable or hardcoded, so we don't have to specify the directory, remote, and branch with every use

It would be funny to put `rustc-subtree` in its own repo and include it here as a subtree repo. But sadly I can't think of a reason, why it has to be in the Rust repo 😄

**Comment by Manishearth** [comment]
I was assuming this would be another rustup component

**Comment by Manishearth** [comment]
Though I guess this doesn't actually need to be versioned with Rust so a cargo install makes more sense

**Comment by oli-obk** [comment]
I marked this as "looking for participation". Working on this git subcommand that can be installed via `cargo install` requires zero work on rustc. Basically what you need to do is to grab the latest version of the subcommand and copy the file with the appropriate license mentioned into a new cargo project. The main function can then `include_str!` the script and invoke `sh` to execute the text (maybe piped in via `stdin`?) by forwarding all command line args to the script. You'll need to add `rust-lang-owner` to the owners of your git repository and the `crates.io` owners of your crate.

Ideally you test the crate by talking with the [clippy team on zulip](https://rust-lang.zulipchat.com/#narrow/stream/257328-clippy) and doing a clippy sync with the "new" tool. If all that works, we'll update the documentation on our `git subtree` usage to reflect the fact that ppl need to install the tool and invoke a slightly different subcommand name.

**Comment by mominul** [comment]
I'd like to work on this issue!

**Comment by oli-obk** [comment]
great! If you have any questions, feel free to ping me on zulip or just ask here. There's also a bot, so if you write `@rustbot claim` it should assign this issue to you.

**Comment by pietroalbini** [comment]
I'm wondering if we need this to actually be on crates.io. We could vendor the script in the repository and call `./x.py subtree`.

**Comment by oli-obk** [comment]
I like having it as an independent thing. Others may want to use it, too. It's not rustc specific

**Comment by mominul** [comment]
Here's [the `gitree` crate](https://crates.io/crates/gitree) which wraps the `git subtree` script. I have included the script form the [git PR](https://github.com/gitgitgadget/git/pull/493) mentioned here. I am not well versed in handling git, so can anyone test it or give me the instructions for doing it? Here's the [Zulip topic](https://rust-lang.zulipchat.com/#narrow/stream/257328-clippy/topic/Issue.20.2370651.20.3A.20A.20Rust.20crate.20wrapping.20the.20git-subtree.20script/near/214999058) about this crate.

**Comment by bjorn3** [comment]
The crate metadata doesn't refer to the repo, so posting it here for others trying to find it: https://github.com/mominul/gitree

**Comment by RalfJung** [comment]
Using subtrees can lead to spuriously re-closing issues that were re-opened. This just happened to https://github.com/rust-lang/rust/issues/82080:
> the original commit did a lot of things, among them fixing clippy. The commit description contained "Fixes #82080". Each commit that affects clippy is synced to clippy (creating a new commit ID since all the non-clippy changes are removed from it), and then that is synced back (with the same commit ID as what it got in clippy). So now we got a fresh commit that still has "Fixes #82080" in its description.

I am not sure what to do about this, but we have a notable risk here of losing track of issues by accidentally closing them.

Can we make it so that the commit that is synthesized for clippy doesn't just have its non-clippy changes removed, but also has its commit message "sanitized" to ensure that it will not cause any automatic actions by GitHub?

**Comment by bjorn3** [comment]
> Can we make it so that the commit that is synthesized for clippy doesn't just have its non-clippy changes removed, but also has its commit message "sanitized" to ensure that it will not cause any automatic actions by GitHub?

I think git subtree would then consider it a different commit and try to pull the original conmit every time you sync.

**Comment by flip1995** [comment]
There's the option in subtree to squash-sync the commits to rustc. But that way we would lose the Clippy history in rustc. 

If we want to keep the history of Clippy and keep using subtree, we just have to live with this, I guess...

**Comment by RalfJung** [comment]
> I think git subtree would then consider it a different commit and try to pull the original conmit every time you sync.

Why that, doesn't it have some kind of bookeeping to match up the commits? Relying on commit messages seems rather fragile?

**Comment by bjorn3** [comment]
I assumed it required the commit to be identical where possible. Based on a cursory reading of the source it seems to only require the tree to be the same: https://github.com/gitgitgadget/git/blob/fe2e4819b869725f870cd3ce99f1f8150fe17dc1/contrib/subtree/git-subtree.sh#L598

**Comment by flip1995** [comment]
Copying comments from #82080:

@jplatte wrote:

> > This sounds like a general problem with the subtree approach...
> 
> IMHO this is more of a problem with GitHub issue / PR references in commit messages. It also happens when merging repos, no matter if you use subtree or sth. else, as long as you keep the history you can have super old commits that suddenly affect (or just reference) new and entirely irrelevant issues. My approach has been to always keep these references out of commits, and only have them in PR descriptions and comments.

**Comment by RalfJung** [comment]
> The problem is, that bors includes the PR body in the commit message.

Good point. Not doing that for the "active" messages could also help. (It'd also cut down on notifications, if bors removed @mentions as well.)

**Comment by Manishearth** [comment]
I feel like that's a pretty rare occurrence and unfortunately the tooling doesn't support doing such a transform currently. We could tweak the bors thing though

**Comment by RalfJung** [comment]
Well, it's also something that has a high chance of flying below the radar when it happens. All we know is it happened *at least* once.

**Comment by tmandry** [comment]
FYI there's been some [discussion of experimenting with miri on Fuchsia](https://bugs.fuchsia.dev/p/fuchsia/issues/detail?id=88691), and transitioning miri to an always-green subtree is one of the primary blockers.

**Comment by oli-obk** [comment]
Clippy is now using subtree, miri is using https://github.com/josh-project/josh/ (a Rust thing that is better and more powerful in various ways than subtree)
